src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2236,60 +2237,53 @@ public Type visitErrorType(ErrorType t, Symbol sym) {
-     * with the given symbol.  If none exists, return null.
+     * Return the base type of t or any of its outer types (by following the types enclosing type)
+     * that start with the given symbol.
+     * If none exists, return null.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asOuterSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                t = t.getEnclosingType();
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
-        }
+        return asOuter(t, sym, Type::getEnclosingType);
     }
 
     /**
-     * Return the base type of t or any of its enclosing types that
-     * starts with the given symbol.  If none exists, return null.
+     * Return the base type of t or any of its outer types (by following owner's
+     * enclosing class) that start with the given symbol.
+     * If none exists, return null.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asEnclosingSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                Type outer = t.getEnclosingType();
-                t = (outer.hasTag(CLASS)) ? outer :
-                    (t.tsym.owner.enclClass() != null) ? t.tsym.owner.enclClass().type :
-                    Type.noType;
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
+        return asOuter(t, sym, type -> getOwnerEnclosingClassType(type));
+    }
+
+    private static Type getOwnerEnclosingClassType(Type type) {
+        return (type.tsym.owner.enclClass() != null)
+                ? type.tsym.owner.enclClass().type
+                : Type.noType;
+    }
+
+    /**
+     * Return the (most specific) base type of t that starts with
+     * the given symbol. If it exists, the base type is returned.
+     * Otherwise, a traversal of outer types attempts to find such
+     * a base type. The traversal of the next outwards type is left
+     * abstract, via the parameter nextType. If none exists, return null.
+     *
+     * @param t a type
+     * @param sym a symbol
+     * @param nextType a unary operator that emits the next type to be examined
+     */
+    public Type asOuter(Type t, Symbol sym, UnaryOperator<Type> nextType) {
```

Comment:
IMHO, this method should be called `asSuperXYZ` -- because that's what it does. It basically calls `asSuper` repeatedly (on the types returned by the unary operator) until a match is found.

Reply:
I agree `asSuperClosure` is nice. IMHO, use that, and rename the other to `supertypeClosure` as (1) that's what it is and (2) it only has one usage that I can see, so renaming is not a big deal and (3) it's a private method in Types, so no risk of it being exposed and accessed elsewhere.